### DISCUSSION TITLE
building all (lnm, atools) in Qt: "No rule to make target" atools from lnm

### DISCUSSION THREAD

**Discussion by u-an-i** [albar965/littlenavmap#692] [Q&A] [answered]
Qt 5.12.11, Windows, MinGW, release:

`:-1: Fehler: No rule to make target 'F:/OneDrive/Dokumente/GitHub/littlenavmap/../build-atools-release/libatools.a', needed by 'littlenavmap.exe'.  Stop.`

created folder was `build-atools-Desktop_Qt_5_12_11_MinGW_64_bit-Release`

files inside that are flat, atools .pro implies a /lib subfolder at least.

lnm .pro implies a build-atools-CONF_TYPE thus i wonder why the ...-atools-Desktop-... folder was created.

**Comment by u-an-i** [comment]
btw.i provided the atools path in the variable, so -CONF_TYPE should not get triggered.

Now checking

**Comment by u-an-i** [comment]
apparent solution:

`ATOOLS_LIB_PATH="F:\\OneDrive\\Dokumente\\GitHub\\build-atools-Desktop_Qt_5_12_11_MinGW_64_bit-Release\\"`

instead of

`ATOOLS_LIB_PATH=F:\OneDrive\Dokumente\GitHub\build-atools-Desktop_Qt_5_12_11_MinGW_64_bit-Release\`

also for all other paths.

now build progessing.

But this results in a qmake error about git not being able to "parse-rev" or similar. But build still started. I wonder why git is attempted, i have not configured it inside qt and i see no .pro entry about git.

**Comment by u-an-i** [comment]
according to internet, other possible issue for the error:

- space(s) in path(s) -> move
- qt got updated and messed up paths -> reinstall qt
- order of project references in .pro -> make sure dependent one are before
- shadow buil is on -> off

**Comment by u-an-i** [comment]
build progressed, deploy progressed, no errors, no output. No deploy folder. just the build folder with all files flat and an .exe.

**Comment by u-an-i** [comment]
i miss "Create and run the target `deploy`" from BUILD.txt

Qt "Deploy project" it is not -- does nothing expect some messages "done in 4s"
Qt "create all configurations" it is not -- creates debug et al

Qt options available in menu in menu order:

build project
create all configurations
"build for run" configuration
run qmake
create compilation database
deploy

I try "build for run"

**Comment by u-an-i** [comment, marked as answer]
no success.

![image](https://user-images.githubusercontent.com/84718885/121800289-4ea78f80-cc31-11eb-9083-531a7432b303.png)


is not reflected in Qt Creator gui.

 build targets in .pro (deploy, copyfiles, all) are not reflected in Qt Creator gui.


Windows has no "copydata" target !  deploy depends on all, all depends on copydata but copydata exists only for linux and mac.

I wonder how "building" from menu did anything at all, assuming qt takes "all" for building because i don't see any mapping done in .pro, and "all" has no commands of its own but only depends on copyall and that does not exits for win, "buildung" in Qt should do nothing -- but it does. "Deployment", which again as assuming automatically takes a target "deploy" because there is no mapping to qt menu commands in .pro, should (try) copy files -- but does nothing.

peculiar.

**Reply by u-an-i** [reply]
regarding configs instead i'm presented debug, profile, release

**Reply by albar965** [reply]
There is no deploy command in atools. The .lib or .a is taken from the build folder by the other projects. Same for headers.

Spaces in paths are quite cumbersome to handle in Windows and qmake. Needs a lot of quotes of quotes. :slightly_frowning_face: 

Building with MinGW saves me the redist installation mess which caused a lot of support requests. No MSVC redist needed now.

**Reply by u-an-i** [reply]
marble is built and installed, atools is built, lnm is built.

How do i get lnm installed / deployed? I currently only have the lnm built folder with all files inside flat without subfolders.

**Reply by albar965** [reply]
Sounds good.

> How do i get lnm installed / deployed? I currently only have the lnm built folder with all files inside flat without subfolders.

`make deploy` in `build-littlenavmap-debug` adds a folder `../deploy/Little Navmap` with all needed files and libraries.

**Reply by u-an-i** [reply]
thank you!!

**Reply by u-an-i** [reply]
run far, do you not mind another issue:
```
F:\OneDrive\Dokumente\GitHub\littlenavmap\etc\SimConnect.dll
1 Datei(en) kopiert
Datei libcrypto-1_1.dll nicht gefunden
0 Datei(en) kopiert
mingw32-make: *** [Makefile:1236: deploy] Error 4
PS F:\OneDrive\Dokumente\GitHub\build-littlenavmap-Desktop_Qt_5_12_11_MinGW_64_bit-Release>
```

on my whole pc, libcrypto-1_1.dll is only inside OneDrive's app folder and i have the openssl repo built and installed.

**Reply by u-an-i** [reply]
i have libcrypto-1_1-x64.dll in lnm folder i copied from openssl i just found :| .

**Reply by u-an-i** [reply]
i now let build the 32 bit open ssl dll.

hint for anyone having built 64 bit before and following the instructions from open ssl rep and getting linker conflict error: use nmake clean before configuring 32 bite

**Reply by u-an-i** [reply]
deploy succesfull, now looking into the zoom issue. Thank you!

**Reply by KOKAProduktion** [reply]
Sry, i forgot about the OpenSSL step... I've added it to the original comment https://github.com/albar965/littlenavmap/discussions/688#discussioncomment-867563

**Comment by u-an-i** [comment]
anyone see also https://github.com/albar965/littlenavmap/discussions/688#discussioncomment-867563

**Reply by albar965** [reply]
> anyone see also [#688 (reply in thread)](https://github.com/albar965/littlenavmap/discussions/688#discussioncomment-867563)

Sure. Very helpful. You mind if I put this into the wiki?

**Reply by u-an-i** [reply]
i don't mind

**Reply by KOKAProduktion** [reply]
Me neither, sure. You might want to doublecheck the Qt and MinGW versions, since you're at the very heartbeat. I've just added another reference to https://github.com/albar965/littlenavmap/discussions/687#discussion-3408464

**Reply by albar965** [reply]
Here you go: https://github.com/albar965/littlenavmap/wiki/CompilingMSVC

**Reply by KOKAProduktion** [reply]
Cool, thx for the credit! But its still a MinGW 32-bit build... Isn't it? VS is just there for the `cmake` command, or did i miss something?

**Reply by albar965** [reply]
> Cool, thx for the credit! But its still a MinGW 32-bit build... Isn't it? VS is just there for the `cmake` command, or did i miss something?

Meh. Very confused now. Because of:

> Host: Win10 64-bit, x64, Visual Studio 2017

Why not use the cmake GUI from https://cmake.org ?

**Reply by KOKAProduktion** [reply]
Sry, didn't mean to cause any confusion. I just had VS already installed, that's why i've mentioned it. I also recall reading about VS in the Marble installation docs... Anyway, i guess cmake.org would have also done the trick.

As @u-an-i discovered, i've forgotten about the OpenSSL step. I've added it to the original comment https://github.com/albar965/littlenavmap/discussions/688#discussioncomment-867563

**Reply by albar965** [reply]
Hope you downloaded the OpenSSL [here](https://bintray.com/vszakats/generic/openssl) and did not try to compile it yourself.

About MSVC: Just a question from the slow one (me): You compile with MinGW using the MSVC GUI or not? 
qmake can generate MSVC project/solution files as far as I know.

**Reply by u-an-i** [reply]
that link leads to no binaries (anymore). i did clone the openssl repo. and succesfully compiled 32 bit and 64 bit now. i set the path in .pro. upon start of lnm i receive warnings about ssl and tls not working but the webserver "works", which is the important thing for me with the built lnm :)

**Reply by albar965** [reply]
:slightly_frowning_face:  Have to look for another source now or compile on our own. Fortunately this affects only Windows.

**Reply by KOKAProduktion** [reply]
Maybe here: https://slproweb.com/products/Win32OpenSSL.html

**Reply by KOKAProduktion** [reply]
> About MSVC: Just a question from the slow one (me): You compile with MinGW using the MSVC GUI or not?
> qmake can generate MSVC project/solution files as far as I know.

Nope, MSVC is just installed on the host and its cmake command is available through PATH.

Which means I only have compiled Marble with it. (cli cmake not MSVC GUI) 

The rest is all qmake & MinGW
